File: internal/core/reboot_story.go
package core

import (
	"context"
	"errors"
	"fmt"
	"os/exec"
	"sort"
	"strings"
	"time"
)

const (
	rebootContextWindow          = 15 * time.Minute
	rebootContextLimit           = 100
	previousBootJournalByteLimit = 16 * 1024
)

type PreviousBootEvidence struct {
	Status     string `json:"status"`
	Evidence   string `json:"evidence,omitempty"`
	Assessment string `json:"assessment"`
}

type RebootStory struct {
	CapturedAt       time.Time             `json:"captured_at"`
	Mode             string                `json:"mode,omitempty"`
	BootID           string                `json:"boot_id,omitempty"`
	BootStartedAt    time.Time             `json:"boot_started_at,omitempty"`
	WindowStart      time.Time             `json:"window_start,omitempty"`
	WindowEnd        time.Time             `json:"window_end,omitempty"`
	PreviousBoot     PreviousBootEvidence  `json:"previous_boot"`
	FailedServices   []ServiceInfo         `json:"failed_services,omitempty"`
	ProblemEvents    []Event               `json:"problem_events,omitempty"`
	RelatedEvents    []Event               `json:"related_events,omitempty"`
	Checks           []Check               `json:"checks"`
	Conclusion       string                `json:"conclusion"`
	Confidence       string                `json:"confidence"`
	CauseAssessment  string                `json:"cause_assessment"`
	ContextNote      string                `json:"context_note"`
}

var previousBootJournalLookup = func(ctx context.Context) (boundedCommandResult, error) {
	lookupCtx, cancel := context.WithTimeout(ctx, 2*time.Second)
	defer cancel()
	command := resolveCommand("journalctl", "/usr/bin/journalctl", "/bin/journalctl")
	return runBoundedCommand(lookupCtx, previousBootJournalByteLimit, command, "-b", "-1", "-n", "24", "--no-pager", "--quiet", "-o", "short-iso")
}

func BuildRebootStory(ctx context.Context, snap Snapshot, events []Event) RebootStory {
	bootID := strings.TrimSpace(snap.Host.BootID)
	bootStartedAt := snap.Host.BootStartedAt
	if bootID == "" || bootStartedAt.IsZero() {
		liveID, liveStartedAt := collectBootIdentity()
		if bootID == "" {
			bootID = liveID
		}
		if bootStartedAt.IsZero() {
			bootStartedAt = liveStartedAt
		}
	}

	story := RebootStory{
		CapturedAt:      time.Now().UTC(),
		Mode:            snap.Mode,
		BootID:          bootID,
		BootStartedAt:   bootStartedAt,
		CauseAssessment: "no reboot cause is claimed from the available evidence",
		ContextNote:     "Boot-time proximity and retained change events are context, not proof of reboot cause. Missing retained evidence does not prove that every service recovered.",
	}
	if !bootStartedAt.IsZero() {
		story.WindowStart = bootStartedAt.Add(-rebootContextWindow)
		story.WindowEnd = bootStartedAt.Add(rebootContextWindow)
		story.RelatedEvents = EventsAround(events, bootStartedAt, rebootContextWindow, rebootContextLimit)
		story.ProblemEvents = postBootProblemEvents(story.RelatedEvents, bootStartedAt)
	}
	story.FailedServices = currentFailedServices(snap.Services)
	story.PreviousBoot = collectPreviousBootEvidence(ctx)
	story.Checks = rebootStoryChecks(story)
	story.Conclusion, story.Confidence = rebootStoryOutcome(story)
	return story
}

func collectPreviousBootEvidence(ctx context.Context) PreviousBootEvidence {
	result, err := previousBootJournalLookup(ctx)
	journal := sanitizeJournal(result.Output)
	assessment := "shutdown classification is unknown; HostSleuth does not infer orderly or abnormal shutdown from missing or bounded journal evidence"
	if errors.Is(err, exec.ErrNotFound) {
		return PreviousBootEvidence{Status: "unknown", Assessment: assessment, Evidence: "journalctl command is unavailable"}
	}
	if err != nil && journal == "" {
		return PreviousBootEvidence{Status: "unknown", Assessment: assessment, Evidence: "previous-boot journal is unavailable: " + boundedEvidence(err.Error(), 384)}
	}
	if journal == "" {
		return PreviousBootEvidence{Status: "unknown", Assessment: assessment, Evidence: "no readable previous-boot journal lines were returned"}
	}
	evidence := boundedEvidence(journal, 4096)
	if result.Truncated {
		evidence += " [previous-boot journal capture truncated at 16 KiB]"
	}
	return PreviousBootEvidence{Status: "available", Assessment: assessment, Evidence: evidence}
}

func currentFailedServices(services []ServiceInfo) []ServiceInfo {
	var out []ServiceInfo
	for _, service := range services {
		if strings.EqualFold(service.Active, "failed") || strings.EqualFold(service.Sub, "failed") {
			out = append(out, service)
		}
	}
	sort.Slice(out, func(i, j int) bool { return out[i].Name < out[j].Name })
	if len(out) > 16 {
		out = out[:16]
	}
	return out
}

func postBootProblemEvents(events []Event, bootStartedAt time.Time) []Event {
	if bootStartedAt.IsZero() {
		return nil
	}
	var out []Event
	for _, event := range events {
		if event.At.Before(bootStartedAt) || !strings.EqualFold(event.Severity, "warning") {
			continue
		}
		switch event.Category {
		case "service", "listener", "container":
			out = append(out, event)
		}
	}
	if len(out) > 32 {
		out = out[:32]
	}
	return out
}

func rebootStoryChecks(story RebootStory) []Check {
	var checks []Check
	if story.BootID == "" && story.BootStartedAt.IsZero() {
		checks = append(checks, Check{Name: "boot-identity", Status: "unknown", Evidence: "kernel boot identity and start time are unavailable"})
	} else {
		evidence := fmt.Sprintf("boot_id=%s", fallback(story.BootID, "unknown"))
		if !story.BootStartedAt.IsZero() {
			evidence += " started_at=" + story.BootStartedAt.Format(time.RFC3339)
		}
		checks = append(checks, Check{Name: "boot-identity", Status: "pass", Evidence: evidence})
	}

	previousStatus := "unknown"
	if story.PreviousBoot.Status == "available" {
		previousStatus = "pass"
	}
	checks = append(checks, Check{Name: "previous-boot-journal", Status: previousStatus, Evidence: boundedEvidence(story.PreviousBoot.Evidence, 1536)})

	if story.Mode == dockerDeploymentMode {
		checks = append(checks, Check{Name: "post-boot-services", Status: "unknown", Evidence: "native systemd service inventory is unavailable in Docker deployment mode"})
	} else if len(story.FailedServices) > 0 {
		checks = append(checks, Check{Name: "post-boot-services", Status: "fail", Evidence: fmt.Sprintf("%d current systemd service(s) are failed", len(story.FailedServices))})
	} else {
		checks = append(checks, Check{Name: "post-boot-services", Status: "pass", Evidence: "no currently failed systemd services are present in the snapshot"})
	}

	if story.BootStartedAt.IsZero() {
		checks = append(checks, Check{Name: "post-boot-changes", Status: "unknown", Evidence: "boot start time is unavailable, so retained changes cannot be anchored to this boot"})
	} else if len(story.ProblemEvents) > 0 {
		checks = append(checks, Check{Name: "post-boot-changes", Status: "fail", Evidence: fmt.Sprintf("%d warning service/listener/container event(s) were retained after boot inside the bounded window", len(story.ProblemEvents))})
	} else {
		checks = append(checks, Check{Name: "post-boot-changes", Status: "pass", Evidence: "no warning service/listener/container events were retained after boot inside the bounded window"})
	}
	return checks
}

func rebootStoryOutcome(story RebootStory) (string, string) {
	if story.BootStartedAt.IsZero() {
		return "current boot start time is unavailable", "medium"
	}
	prefix := "current boot started " + story.BootStartedAt.Format(time.RFC3339)
	if len(story.FailedServices) > 0 || len(story.ProblemEvents) > 0 {
		return prefix + "; post-boot service/listener/container problems are visible in current or retained evidence", "high"
	}
	return prefix + "; no retained post-boot service/listener/container problem evidence was found", "medium"
}

File: cmd/hostsleuth/web/reboot_story.js
function installRebootStoryUI() {
  const changes = byId("view-changes");
  if (!changes || byId("rebootStoryPanel")) return;

  changes.insertAdjacentHTML("beforeend", `
    <section class="panel reboot-story-panel" id="rebootStoryPanel">
      <div class="reboot-story-heading">
        <div>
          <p class="eyebrow">REBOOT STORY</p>
          <h2>What happened around this boot—and what failed to come back?</h2>
          <p class="section-copy">Correlate kernel boot identity, bounded previous-boot journal evidence when readable, current failed services, and retained changes near boot. HostSleuth does not invent a reboot cause.</p>
        </div>
        <button class="primary-button" id="rebootStoryButton" type="button">Build reboot story</button>
      </div>
    </section>

    <section class="panel diagnosis-result hidden" id="rebootStoryResult" aria-live="polite">
      <div class="diagnosis-heading">
        <div>
          <p class="eyebrow">BOOT / RECOVERY ANSWER</p>
          <h2 id="rebootStoryConclusion">Reboot story complete</h2>
          <p class="muted" id="rebootStoryBootMeta"></p>
          <p class="section-copy" id="rebootStoryCause"></p>
          <p class="context-note" id="rebootStoryContext"></p>
        </div>
        <span class="confidence-badge" id="rebootStoryConfidence"></span>
      </div>
      <div class="reboot-story-grid">
        <div class="context-card">
          <p class="subsection-label">Current recovery state</p>
          <div id="rebootFailedServices"></div>
        </div>
        <div class="context-card">
          <p class="subsection-label">Previous boot evidence</p>
          <div id="rebootPreviousBoot"></div>
        </div>
      </div>
      <div class="diagnosis-context-grid reboot-story-events-grid">
        <div>
          <p class="subsection-label">Post-boot problems in bounded window</p>
          <div id="rebootProblemEvents" class="mini-event-list"></div>
        </div>
        <aside class="context-card">
          <p class="subsection-label">All retained changes near boot</p>
          <div id="rebootRelatedEvents" class="mini-event-list"></div>
        </aside>
      </div>
    </section>

    <section class="panel diagnosis-result hidden" id="rebootStoryError" aria-live="polite">
      <p class="eyebrow">REBOOT STORY ERROR</p>
      <h2>HostSleuth could not build the reboot story.</h2>
      <p class="section-copy" id="rebootStoryErrorText"></p>
    </section>
  `);

  byId("rebootStoryButton").addEventListener("click", runRebootStory);
}

function renderRebootEventList(container, events, emptyText) {
  container.replaceChildren();
  if (!events?.length) {
    container.append(makeEmpty(emptyText));
    return;
  }
  events.forEach((event) => {
    const item = document.createElement("div");
    item.className = "mini-event";
    const summary = document.createElement("strong");
    summary.textContent = event.summary || "Recorded change";
    const meta = document.createElement("span");
    meta.textContent = `${event.category || "change"} · ${formatTime(event.at)}`;
    item.append(summary, meta);
    container.append(item);
  });
}

function renderRebootStory(story) {
  byId("rebootStoryError").classList.add("hidden");
  byId("rebootStoryResult").classList.remove("hidden");
  text(byId("rebootStoryConclusion"), story.conclusion || "Reboot story complete");
  text(byId("rebootStoryConfidence"), `${story.confidence || "unknown"} confidence`);
  const id = story.boot_id ? ` · boot ${story.boot_id.slice(0, 12)}` : "";
  text(byId("rebootStoryBootMeta"), `${story.boot_started_at ? formatTime(story.boot_started_at) : "boot start unavailable"}${id}`);
  text(byId("rebootStoryCause"), story.cause_assessment || "No reboot cause is claimed.");
  text(byId("rebootStoryContext"), story.context_note || "Boot-time proximity is context, not causation.");

  const failed = byId("rebootFailedServices");
  failed.replaceChildren();
  if (!story.failed_services?.length) {
    failed.append(makeEmpty(story.mode === "docker" ? "Native systemd service inventory is unavailable in Docker mode." : "No currently failed systemd services are present in the snapshot."));
  } else {
    story.failed_services.forEach((service) => {
      const item = document.createElement("div");
      item.className = "reboot-service-row";
      const name = document.createElement("strong");
      name.textContent = service.name || "service";
      const state = document.createElement("span");
      state.textContent = `${service.active || "unknown"}/${service.sub || "unknown"}`;
      item.append(name, state);
      failed.append(item);
    });
  }

  const previous = byId("rebootPreviousBoot");
  previous.replaceChildren();
  const status = document.createElement("strong");
  status.textContent = story.previous_boot?.status === "available" ? "Previous-boot journal available" : "Previous-boot journal unavailable / unknown";
  const assessment = document.createElement("p");
  assessment.className = "context-note";
  assessment.textContent = story.previous_boot?.assessment || "Shutdown classification is unknown.";
  previous.append(status, assessment);
  if (story.previous_boot?.evidence) {
    const details = document.createElement("details");
    const summary = document.createElement("summary");
    summary.textContent = "Show bounded evidence";
    const evidence = document.createElement("pre");
    evidence.className = "reboot-journal-evidence";
    evidence.textContent = story.previous_boot.evidence;
    details.append(summary, evidence);
    previous.append(details);
  }

  renderRebootEventList(byId("rebootProblemEvents"), story.problem_events || [], "No warning service/listener/container events were retained after boot inside the bounded window.");
  renderRebootEventList(byId("rebootRelatedEvents"), story.related_events || [], "No retained host changes were found inside the boot window.");
}

async function runRebootStory() {
  const button = byId("rebootStoryButton");
  const result = byId("rebootStoryResult");
  const errorPanel = byId("rebootStoryError");
  button.disabled = true;
  button.textContent = "Building story…";
  result.classList.add("hidden");
  errorPanel.classList.add("hidden");
  try {
    const response = await fetch("/api/reboot-story", { cache: "no-store" });
    if (!response.ok) throw new Error((await response.text()) || `request failed (${response.status})`);
    renderRebootStory(await response.json());
  } catch (error) {
    text(byId("rebootStoryErrorText"), error.message || "Unknown error");
    errorPanel.classList.remove("hidden");
  } finally {
    button.disabled = false;
    button.textContent = "Build reboot story";
  }
}

installRebootStoryUI();
